tools/ansible/roles/dockerfile/templates/Dockerfile.j2

```diff
@@ -38,15 +38,22 @@ RUN dnf -y update && dnf install -y 'dnf-command(config-manager)' && \
     patch \
     postgresql \
     postgresql-devel \
-    python3-devel \
-    python3-pip \
-    python3-setuptools \
+    python3.11 \
+    "python3.11-devel" \
+    "python3.11-pip" \
+    "python3.11-setuptools" \
+    "python3.11-packaging" \
+    "python3.11-psycopg2" \
     swig \
     unzip \
     xmlsec1-devel \
     xmlsec1-openssl-devel
 
-RUN pip3 install virtualenv build psycopg
+RUN update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.11 1
```

Comment:
na. I would not use update-alternatives here. It's not portable. Also, this makes changes in a lot more places than just referring to the python3.11 bin directly. I'd rather see a wrapper script than update-alternatives. But, even that is overkill.

Reply:
@jjwatt 

I'm confused by your feedback and I would love to learn more.
You indicate that update-alternatives "is not portable".
However, we're building a docker image `tools/ansible/roles/dockerfile/templates/Dockerfile.j2` from `FROM quay.io/centos/centos:stream9`, it's provided explicitly in the base rpms as this is the expected way to manage multiple versions in linux.

By hard coding python versions around the place (current approach), aren't we making headaches for ourselves.
By changing the default version in the system, this is "catch all" for any python process instantiated.
I would have thought that this method is the "few changes as possible" approach, in that rather than having python 3.11 hardcoded all over build scripts and other locations, we just change it in 1 place, hence 1 change.  

When we install python 3.12 in the future, we only have to run update-alternatives to switch to the new version.  All system administrators are accustomed to update-alternatives and "xx --version". 
Fedora/rhel/centos changes over time and versions are expected to change, so why would you expect a python version to remain the same every time?
python 3.9 is expected to EOL at end of 2024
https://devguide.python.org/versions/

But as I said, I would love to learn more and your experience from this.
I would also love to not have this task on our EPIC list in the future.
Having hard coded version dependencies in itself is commonly referred to as "bad code smell".
We should avoid as much as possible.

Reply:
> it's provided explicitly in the base rpms as this is the expected way to manage multiple versions in linux.
slight correction: on RedHat/CentOS (taken from Debian, so it's there, too) but it's not in most Linux distros or in most container images (e.g., Alpine, distroless, etc).

NOTE: We did not always use the centos9 stream container. Couldn't that change again?

> By hard coding python versions around the place (current approach), aren't we making headaches for ourselves.
By changing the default version in the system, this is "catch all" for any python process instantiated.
I would have thought that this method is the "few changes as possible" approach, in that rather than having python 3.11 hardcoded all over build scripts and other locations, we just change it in 1 place, hence 1 change.

Valid point, but you should list the headaches! only one I see is writing the code and changing the py version in every place. this is good. it's a huge deal to change the python version and you shouldn't do it often.

NOTE: Python, its entire lib and dependency stack varies on these axes:
Major.Minor version, Operating System, Platform

We don't want to *change* the systems' python! that could break all kinds of things. Python happily breaks their parsers between Minor versions!

The issues come in from non-determinism:

you update alternatives and CentOS wants to run something for the system. That was tested on the SYSTEM python it was integrated with. You can break a lot this way, in ways we cannot predict. AND it would be super hard to debug, imho. We don't NEED to change the system python, so why would we create more vairance and upredictability where we don't NEED it? Writing code and changing the hard-coded values aren't really "headaches." We can decrease the callsites (where changes have to be made) in other ways that won't effect the SYSTEM and any other packages, right?

> Having hard coded version dependencies in itself is commonly referred to as "bad code smell".
We should avoid as much as possible.
Maybe. But it's a GOOD BUILD SYSTEM smell ;)

Reply:
went with hard coded 3.11 and also changed a few places to make it more specific from @jjwatt feedback.
Also, we hardened pip3 install from a git repo by specifying a sha1.